todo_app/app.py

```diff
@@ -13,59 +17,167 @@ def create_app():
     mongo_db_client = pymongo.MongoClient(os.getenv('DATABASE_CONNECTION_STRING'))
     db = mongo_db_client[os.getenv('DATABASE_NAME')]
     collection = db['todo_app_items']
+    users_collection = db['users']
 
     app = Flask(__name__)
     app.config.from_object('todo_app.flask_config.Config')
+    app.secret_key = os.getenv('SECRET_KEY')
+    login_manager = LoginManager()
+
+    web_application_client = WebApplicationClient(os.getenv('CLIENT_ID'))
+
+    class User(UserMixin):
+        def __init__(self, id):
+            self.id = id
+
+    @login_manager.unauthorized_handler
+    def unauthenticated():
+        return redirect(web_application_client.prepare_request_uri('https://github.com/login/oauth/authorize'))
+    
+    @login_manager.user_loader     
+    def load_user(user_id):         
+        user = User(user_id)
+        query = {'github_id': user_id}
+        db_user = users_collection.find_one(query)
+        user.name = db_user['name']
+        user.user_role = db_user['user_role']
+        return user
+    
+    login_manager.init_app(app)
+
+    @app.route('/login/callback', methods=['GET'])
+    def login_user_callback():    
+        token_url, headers, data = web_application_client.prepare_token_request(
+            'https://github.com/login/oauth/access_token', code = request.args['code'], 
+            client_id = os.getenv('CLIENT_ID'), client_secret = os.getenv('CLIENT_SECRET')
+        )
+        
+        headers['Accept'] = 'application/json'
+        response = requests.post(token_url, headers=headers, data=data)
+        web_application_client.parse_request_body_response(response.text)
+        access_token = web_application_client.token['access_token']
+        header = {'Authorization': f'Bearer {access_token}'}
+        response = requests.get('https://api.github.com/user', headers=header).json()
+        github_id = str(response['id'])
+        
+        query = {'github_id': github_id}
+        db_user = users_collection.find_one(query)
+        
+        if(db_user) == None:
+            if(users_collection.count_documents({}) == 0):
+                user_role = 'admin'
+            else:
+                user_role = 'reader'
+            
+            post = {
+                'name': response['login'],
+                'github_id': github_id,
+                'user_role': user_role
+            }
+            users_collection.insert_one(post)
+            db_user = users_collection.find_one(query)
+
+        user = User(db_user['github_id'])
+        login_user(user)
+        return redirect('/')
+
+    def is_writer():
+        if(app.config['LOGIN_DISABLED']):
+            return True
+        return current_user.user_role == 'writer'
```

Comment:
Can also be written as `return app.config['LOGIN_DISABLED'] or current_user.user_role == 'writer'`